Merge workflows

jneira opened the pull request:
---
Please include the following checklist in your PR:

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

robx committed:
Bootstrap generation fix for exe-only dependencies

This isn't currently a concrete problem, but surfaced when adding
a dependency to cabal-install:exe.

jneira committed:
Correct typo in validate.sh

jneira committed:
Merge all os's

jneira committed:
Add windows arch to validate.sh

jneira committed:
Dont use cli for win and 8.0.2

jneira committed:
Remove compile warnings

jneira committed:
temporary no fail-fast

jneira committed:
Install autoconf in mac

jneira committed:
Use full path to ghc

jneira committed:
Dont execute hackage tests in win

jneira committed:
Make shorter memory tests suite name

andreabedini commented on .github/workflows/validate.yml:
```diff
@@ -16,50 +20,52 @@ on:
+      fail-fast: false
       matrix:
+        os: ["ubuntu-latest", "macos-latest", "windows-latest"]
+        ghc: ["9.2.1", "9.0.2", "9.0.1", "8.10.7", "8.8.4", "8.6.5", "8.4.4", "8.0.2"]
```
I think you need to add `experimental: ["false"]` otherwise `experimental` might not be defined for all combinations.

jneira replied:
yeah, as  conditions are over the value true, if it is not defined it is not experimental

andreabedini commented on .github/workflows/validate.yml:
```diff
@@ -16,50 +20,52 @@ on:
+            experimental: "true" # see #7988 and #7989
+          - os: "windows-latest"
+            ghc: "8.0.2"
+            cli: "false"
```
I don't understand how you are building this matrix. Do you want to list cases one by one or do a full cross-product, or a combination of both? Eg. ubuntu-latest, ghc 8.0.2 is included twice, once with `cli: false` and the other with `cli` undefined (AFAIU)

jneira replied:
afaik if you repeat params from the product it only adds the new param, you can check in the workflow run there are no repeated jobs

jneira replied:
it is less explicit but it allows reduce concrete jobs list

andreabedini commented on .github/workflows/validate.yml:
```diff
@@ -90,43 +96,69 @@ jobs:
 
       - name: Validate print-tool-versions
-        run: sh validate.sh -j 2 -w ghc-${{ matrix.ghc }} -v ${{ matrix.flags }} -s print-tool-versions
+        run: sh validate.sh -j 2 -w ghc-${{ matrix.ghc }} -v ${{ steps.validate-inputs.outputs.flags }} -s print-tool-versions
```
Do we mean to pass flags as `$FLAGS` now?

jneira replied:
No cause $FLAGS in the previous step is bound to the outputs of such step and $FLAGS is not visible directly from out the step is using it. That is a nice way to hide the details of how a step computes data imo.

jneira replied:
i can imagine we could set validate targets for example in that step

andreabedini commented on .github/workflows/validate.yml:
```diff
@@ -90,43 +96,69 @@ jobs:
       - name: Validate cli-tests
-        if: matrix.cli == true
-        run: sh validate.sh -j 2 -w ghc-${{ matrix.ghc }} -v ${{ matrix.flags }} -s cli-tests
+        if: matrix.cli != 'false'
```
Is this because `matrix.cli` is not defined everywhere?

jneira replied:
yeah

andreabedini commented on cabal-install/cabal-install.cabal:
```diff
@@ -332,7 +332,7 @@ Test-Suite unit-tests
 -- Tests to run with a limited stack and heap size
 --
-Test-Suite memory-usage-tests
+Test-Suite mem-use-tests
```
Why this change?

jneira replied:
cause original name make the build generates paths which exceeds windows max path limit (256) and the build fail for some ghc versions (<= 8.6.5)

jneira replied:
will add a comment

andreabedini reviewed (commented):
I left some comments

jneira committed:
Add comment about the name change

warning about the windows max path issue

jneira committed:
Use FLAGS as global env var

jneira committed:
Fix abs paths test for windows

jneira committed:
MArk unit tests broken for win

jneira committed:
Mark ForeignLibs broken for win and 9.0

Referencing issue #7989

jneira commented:
cli tests in windows using bash are failing due to max path issues, they dont fail using powershell as shell though:

* https://github.com/jneira/cabal/runs/5433647403?check_suite_focus=true#step:14:361

```
D:\a\cabal\cabal\dist-val-ghc-8.6.5\build\x86_64-windows\ghc-8.6.5\cabal-testsuite-3\build\cabal-tests\cabal-tests.exe --builddir=/d/a/cabal/cabal/dist-val-ghc-8.6.5/build/x86_64-windows/ghc-8.6.5/cabal-testsuite-3 --with-cabal=D:\a\cabal\cabal\dist-val-ghc-8.6.5\build\x86_64-windows\ghc-8.6.5\cabal-install-3.7.0.0\x\cabal\build\cabal\cabal.exe -j2 --hide-successes
```

```
creating D:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\my-custom-preprocessor-tmp

"C:\ProgramData\chocolatey\lib\ghc.8.6.5\tools\ghc-8.6.5\bin\ghc-8.6.5.exe" "--make" "-no-link" "-fbuilding-cabal-package" "-O" "-static" "-outputdir" "D:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\my-custom-preprocessor-tmp" "-odir" "D:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\my-custom-preprocessor-tmp" "-hidir" "D:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\my-custom-preprocessor-tmp" "-stubdir" "D:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\my-custom-preprocessor-tmp" "-i" "-iD:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\my-custom-preprocessor-tmp" "-i." "-iD:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\autogen" "-iD:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\global-autogen" "-ID:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\autogen" "-ID:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\global-autogen" "-ID:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\my-custom-preprocessor-tmp" "-optP-include" "-optPD:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\autogen\cabal_macros.h" "-hide-all-packages" "-Wmissing-home-modules" "-no-user-package-db" "-package-db" "D:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\home\.cabal\store\ghc-8.6.5\package.db" "-package-db" "D:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\packagedb\ghc-8.6.5" "-package-db" "D:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\package.conf.inplace" "-package-id" "base-4.12.0.0" "-package-id" "directory-1.3.3.0" "-XHaskell2010" ".\MyCustomPreprocessor.hs" "-hide-all-packages"

[1 of 1] Compiling Main             ( MyCustomPreprocessor.hs, D:\\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\my-custom-preprocessor-tmp\Main.o )

Assembler messages:

Fatal error: can't create D:\\\\a\\\\cabal\\\\cabal\\\\cabal-testsuite\\\\PackageTests\\\\CustomPreProcess\\\\cabal.dist\\\\work\\\\.\\\\dist\\\\build\\\\x86_64-windows\\\\ghc-8.6.5\\\\internal-preprocessor-test-0.1.0.0\\\\build\\\\my-custom-preprocessor\\\\my-custom-preprocessor-tmp\\Main.o: No such file or directory

`gcc.exe' failed in phase `Assembler'. (Exit code: 1)

```

* good: https://github.com/jneira/cabal/runs/5428036153?check_suite_focus=true#step:14:45512

```
cabal v2-run cabal-testsuite:cabal-tests -- -j1 -v --with-cabal=dist-newstyle\build\x86_64-windows\ghc-8.6.5\cabal-install-3.7.0.0\x\cabal\build\cabal\cabal.exe
```

```
creating D:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\my-custom-preprocessor-tmp

"C:\ProgramData\chocolatey\lib\ghc.8.6.5\tools\ghc-8.6.5\bin\ghc.exe" "--make" "-no-link" "-fbuilding-cabal-package" "-O" "-static" "-outputdir" "D:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\my-custom-preprocessor-tmp" "-odir" "D:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\my-custom-preprocessor-tmp" "-hidir" "D:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\my-custom-preprocessor-tmp" "-stubdir" "D:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\my-custom-preprocessor-tmp" "-i" "-iD:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\my-custom-preprocessor-tmp" "-i." "-iD:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\autogen" "-iD:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\global-autogen" "-ID:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\autogen" "-ID:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\global-autogen" "-ID:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\my-custom-preprocessor-tmp" "-optP-include" "-optPD:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\autogen\cabal_macros.h" "-hide-all-packages" "-Wmissing-home-modules" "-no-user-package-db" "-package-db" "D:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\home\.cabal\store\ghc-8.6.5\package.db" "-package-db" "D:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\packagedb\ghc-8.6.5" "-package-db" "D:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\.\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\package.conf.inplace" "-package-id" "base-4.12.0.0" "-package-id" "directory-1.3.3.0" "-XHaskell2010" ".\MyCustomPreprocessor.hs" "-hide-all-packages"

[1 of 1] Compiling Main             ( MyCustomPreprocessor.hs, D:\a\cabal\cabal\cabal-testsuite\PackageTests\CustomPreProcess\cabal.dist\work\dist\build\x86_64-windows\ghc-8.6.5\internal-preprocessor-test-0.1.0.0\build\my-custom-preprocessor\my-custom-preprocessor-tmp\Main.o )
```

The unique way i can think off is trying to reduce tests names to make them work in both shells as it is very inconvenient in general being not able to run them in a msys2 bash shell

(Posted in cabal https://github.com/haskell/cabal/issues/6271#issuecomment-1062776762)

jneira committed:
Fix unexpected ok for windows

jneira committed:
Ignore windows test artifacts

jneira committed:
Special case for win and old ghcs

jneira committed:
Fix validate workflow

jneira committed:
Clean matrix and fix old ghcs

jneira committed:
Add post job

jneira committed:
Remove superfluous spaces

jneira committed:
Skip T3827 for ghc-9.0.2, see #8032

jneira committed:
upload cabal executable

jneira committed:
Add dogfooding job

jneira committed:
Tar cabal executable

jneira committed:
Use constants and clean up

jneira committed:
Add changelog for #7279 and #8006

jneira committed:
Mark max path tests as broken

jneira committed:
Make it fail if any job is cancelled

jneira committed:
Dont run any hackage tests on win

jneira committed:
Recover bad rebase

Mikolaj committed:
Merge pull request #8041 from jneira/changelog-7279

Add changelog for #7279 and #8006

jneira committed:
Mark more tests as broken

mergify[bot] committed:
Merge pull request #8042 from robx/fix-bootstrap-selfref

Bootstrap generation fix for exe-only dependencies

jneira committed:
Make shorter paths

jneira committed:
Correct ghc bounds

jneira committed:
Ref to #3126 and refactor knownBroken

jneira committed:
Restore fail-fast true

jneira committed:
Use 8.10.7 for exhaustive tests

jneira committed:
Use ghc for release in dogfooding

Using validate job outputs cause it seems
you can't use the env context directly in matrix
